### ISSUE TITLE
'error': 'transpose expects a vector of size 6. But input(1) is a vector of size 3'. Issue with bidirectional compatibility with tensorflow serving with batch/sample number dimension.

### ISSUE THREAD

**Issue by bigbizze** [tensorflow/serving#1269]
* Windows 10 - Version 1809 OS Build 17763.292
* Docker - Version 18.03.0-ce (Client) - Version 18.09.1 (Engine)
* Tensorflow Serving was installed using the command "docker pull tensorflow/serving"
* TensorFlow Serving - Version 1.12.0

Issue with batch dimension when using tensorflow serving...

metadata:
`{'model_spec': {'name': 'prod_mod', 'signature_name': '', 'version': '1'}, 'metadata': {'signature_def': {'signature_def': {'serving_default': {'inputs': {'input': {'dtype': 'DT_FLOAT', 'tensor_shape': {'dim': [{'size': '-1', 'name': ''}, {'size': '50', 'name': ''}], 'unknown_rank': False}, 'name': 'embedding_1_input:0'}}, 'outputs': {'output': {'dtype': 'DT_FLOAT', 'tensor_shape': {'dim': [{'size': '-1', 'name': ''}, {'size': '1', 'name': ''}], 'unknown_rank': False}, 'name': 'output/Sigmoid:0'}}, 'method_name': 'tensorflow/serving/predict'}}}}}`

error response:
```
<Response [400]>
{'error': 'transpose expects a vector of size 6. But input(1) is a vector of size 3\n\t [[{{node bidirectional_1/transpose}} = Transpose[T=DT_FLOAT, Tperm=DT_INT32, _class=["loc:@bidirectional_1/TensorArrayUnstack/TensorArrayScatter/TensorArrayScatterV3"], _output_shapes=[[50,?,300]], _device="/job:localhost/replica:0/task:0/device:CPU:0"](embedding_1/embedding_lookup, Attention/transpose/perm)]]'}
```

I am using Keras to build a model and am exporting it to a .pb file using "tf.saved_model.simple_save".

The input to my model is (num samples, 50) which becomes (none, 50) and when making a POST request to the tensorflow serving instance in Docker if I reshape my input so it is only one sample with the shape (50) I get a response & prediction as expected, but if I attempt to batch more than one sample together it gives me the above error. 

I am using a custom layer in this model which is referenced at the end of the error "Attention/transpose/perm" but what is really confusing me is earlier when it says "_output_shapes=[[50,?,300]] which makes me think that it is moving the num samples to dimension 1 from 0 for some reason.

This is the command I am using to create the Docker tensorflow serving instance: 

`docker run -p 9000:9000 --name tfserve --mount type=bind,source=//c/Users/Charles/PycharmProjects/spamreader/serving/prod_mod,target=/models/prod_mod,target=/models/prod_mod -t --entrypoint=tensorflow_model_server tensorflow/serving --enable_batching --rest_api_port=9000 --model_name=prod_mod --model_base_path=/models/prod_mod
 `

Is this an error of compatibility with Bidirectional layers and tensorflow serving?

I had opened up a similar Issue thread a few weeks ago but closed it after I had thought I had solved the issue but in reality I had just discovered that the prediction works as intended only when their is no sample dimension.

It looks like a similar issue as this one in terms of how it is manifesting even if the surrounding context is quite different: https://github.com/migueldeicaza/TensorFlowSharp/issues/288

**Comment by hgadig** [comment]
Identify the size of the input tensor and change it accordingly as the error suggests.

Please refer[ this](https://stackoverflow.com/questions/51326140/tensorflow-transpose-expects-a-vector-of-size-1-but-input1-is-a-vector-of-si) for more information.

**Comment by bigbizze** [comment]
As I said in my post, when the input is altered so as to remove the batch dimension entirely the model serves as intended. When I try to batch the input it throws this error.

The input data's shape works perfectly as intended with batching when I use the same data and test using Keras' model.predict() function.

**Comment by bigbizze** [comment]
Okay so I think I found an important clue into the problem... My model, when compiled, has an output tensor shape of (None, 1), but when I use Keras' "model.save()" feature to actually save the model the output shape changes for some reason to (None, 50, 1), so the model when inferred upon using tensorflow serving thinks that the batch dimension is actually the input length dimension.

**Comment by bigbizze** [comment]
Here is the model.summary() printout for reference:

`__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
==================================================================================================
input (InputLayer)              (None, 50)           0                                            
__________________________________________________________________________________________________
embedding_1 (Embedding)         (None, 50, 300)      34830600    input[0][0]                      
__________________________________________________________________________________________________
drop1 (Dropout)                 (None, 50, 300)      0           embedding_1[0][0]                
__________________________________________________________________________________________________
conv1d_1 (Conv1D)               (None, 50, 50)       30050       drop1[0][0]                      
__________________________________________________________________________________________________
conv1d_2 (Conv1D)               (None, 50, 50)       5050        conv1d_1[0][0]                   
__________________________________________________________________________________________________
drop2 (Dropout)                 (None, 50, 50)       0           conv1d_2[0][0]                   
__________________________________________________________________________________________________
global_average_pooling1d_1 (Glo (None, 50)           0           drop2[0][0]                      
__________________________________________________________________________________________________
global_max_pooling1d_1 (GlobalM (None, 50)           0           drop2[0][0]                      
__________________________________________________________________________________________________
concatenate_1 (Concatenate)     (None, 100)          0           global_average_pooling1d_1[0][0] 
                                                                 global_max_pooling1d_1[0][0]     
__________________________________________________________________________________________________
denseout1 (Dense)               (None, 256)          25856       concatenate_1[0][0]              
__________________________________________________________________________________________________
activout1 (Activation)          (None, 256)          0           denseout1[0][0]                  
__________________________________________________________________________________________________
dropout1 (Dropout)              (None, 256)          0           activout1[0][0]                  
__________________________________________________________________________________________________
denseout2 (Dense)               (None, 128)          32896       dropout1[0][0]                   
__________________________________________________________________________________________________
activout2 (Activation)          (None, 128)          0           denseout2[0][0]                  
__________________________________________________________________________________________________
output (Dense)                  (None, 1)            129         activout2[0][0]                  
==================================================================================================
Total params: 34,924,581
Trainable params: 93,981
Non-trainable params: 34,830,600`

**Comment by bigbizze** [comment]
Okay so I'm fairly confident that the issue is that I assumed you could send multiple batches as one array to the model to run predictions as batches and in reality I needed to batch each sample of the data in the request payload and then the model will infer in batches.

I.e. my original attempt was:

`payload = {

    "instances": [{"input": out}]
    # "instances": [{'inputs': out2}]
}`

where "out" was an array with the shape (number of samples, input length)

When I actually needed to use:
payload = {

    "instances": [{"input": out},
                  {"input": out},
                   {"input": out}]
}

Where each "out" is a single sample included in the batch which is the request.

**Comment by sathyamoorthyrr** [comment]
When I try to run Batch mode with structure,
```
{ "inputs":
       [
         {"source_ids": [ _some value_ ], "source_len": [ _some value_ ]},
         {"source_ids": [ _some value_ ], "source_len": [ _some value_ ]}
       ]
}
```
I get,

> {
>     "error": "inputs is a plain value/list, but expecting an object as multiple input tensors required as per tensorinfo_map"
> }

**Comment by lilao** [comment]
@bigbizze: thanks for following up with your findings!

@sathyamoorthyrr: your question seems different from the original issue. Can you open a new issue?

**Comment by DL-POWER** [comment]
> * Windows 10 - Version 1809 OS Build 17763.292
> * Docker - Version 18.03.0-ce (Client) - Version 18.09.1 (Engine)
> * Tensorflow Serving was installed using the command "docker pull tensorflow/serving"
> * TensorFlow Serving - Version 1.12.0
> 
> Issue with batch dimension when using tensorflow serving...
> 
> metadata:
> `{'model_spec': {'name': 'prod_mod', 'signature_name': '', 'version': '1'}, 'metadata': {'signature_def': {'signature_def': {'serving_default': {'inputs': {'input': {'dtype': 'DT_FLOAT', 'tensor_shape': {'dim': [{'size': '-1', 'name': ''}, {'size': '50', 'name': ''}], 'unknown_rank': False}, 'name': 'embedding_1_input:0'}}, 'outputs': {'output': {'dtype': 'DT_FLOAT', 'tensor_shape': {'dim': [{'size': '-1', 'name': ''}, {'size': '1', 'name': ''}], 'unknown_rank': False}, 'name': 'output/Sigmoid:0'}}, 'method_name': 'tensorflow/serving/predict'}}}}}`
> 
> error response:
> 
> ```
> <Response [400]>
> {'error': 'transpose expects a vector of size 6. But input(1) is a vector of size 3\n\t [[{{node bidirectional_1/transpose}} = Transpose[T=DT_FLOAT, Tperm=DT_INT32, _class=["loc:@bidirectional_1/TensorArrayUnstack/TensorArrayScatter/TensorArrayScatterV3"], _output_shapes=[[50,?,300]], _device="/job:localhost/replica:0/task:0/device:CPU:0"](embedding_1/embedding_lookup, Attention/transpose/perm)]]'}
> ```
> 
> I am using Keras to build a model and am exporting it to a .pb file using "tf.saved_model.simple_save".
> 
> The input to my model is (num samples, 50) which becomes (none, 50) and when making a POST request to the tensorflow serving instance in Docker if I reshape my input so it is only one sample with the shape (50) I get a response & prediction as expected, but if I attempt to batch more than one sample together it gives me the above error.
> 
> I am using a custom layer in this model which is referenced at the end of the error "Attention/transpose/perm" but what is really confusing me is earlier when it says "_output_shapes=[[50,?,300]] which makes me think that it is moving the num samples to dimension 1 from 0 for some reason.
> 
> This is the command I am using to create the Docker tensorflow serving instance:
> 
> `docker run -p 9000:9000 --name tfserve --mount type=bind,source=//c/Users/Charles/PycharmProjects/spamreader/serving/prod_mod,target=/models/prod_mod,target=/models/prod_mod -t --entrypoint=tensorflow_model_server tensorflow/serving --enable_batching --rest_api_port=9000 --model_name=prod_mod --model_base_path=/models/prod_mod `
> 
> Is this an error of compatibility with Bidirectional layers and tensorflow serving?
> 
> I had opened up a similar Issue thread a few weeks ago but closed it after I had thought I had solved the issue but in reality I had just discovered that the prediction works as intended only when their is no sample dimension.
> 
> It looks like a similar issue as this one in terms of how it is manifesting even if the surrounding context is quite different: [migueldeicaza/TensorFlowSharp#288](https://github.com/migueldeicaza/TensorFlowSharp/issues/288)

Heyy Bro, i was reading this page and i found you wrote about custom layers.I also want to try some custom layers but after lot of research i didn't found anything relevant can u help me with it??

**Comment by anyongjin** [comment]
Hi, I'm facing the same problem as you. I am using Keras to build a model which has an output tensor shape of (None, 1), and exported it to a pb file using tf.saved_model.simple_save, 
I got an Error "transpose expects a vector of size 4. But input(1) is a vector of size 3'" while call the model served on tf serving with grpc, and it works  as intended only when their is no sample dimension.
finally, I found the problem:
```python
                tensor_shape = [1] + list(data.shape)
                dims = [tensor_shape_pb2.TensorShapeProto.Dim(size=dim) for dim in tensor_shape]
                tensor_shape = tensor_shape_pb2.TensorShapeProto(dim=dims)
                tensor = tensor_pb2.TensorProto(
                    dtype=types_pb2.DT_FLOAT,
                    tensor_shape=tensor_shape,
                    float_val=list(data.reshape(-1)))
```
I replaced the *tf.make_tensor_proto* with the code above, and the tensor_shape is limited to one sample
Everything is ok if the first line is replaced with "tensor_shape = list(data.shape)"  
hope helpful for someone...
